### PR TITLE
Add fetch_userinfo to upstream SSO provider

### PR DESCRIPTION
**Pull request by MatMaul**
Untested for now so I am keeping it as a draft.

### CODE REVIEW

File: crates/handlers/src/upstream_oauth2/callback.rs
Comment (sandhose) [inline crates/handlers/src/upstream_oauth2/callback.rs:264]: Just thinking out loud at this point, but it would be nice to have both the userinfo and the idtoken claims in the environment.

I don't know what the right config is to tell whether to fetch userinfo or not for example, *but* I would set `userinfo_claims` in the template environment if we fetched it, `id_token_claims` if we have an `id_token` and `user` being somehow a merge of the two?

One stretch goal of this would be for this to work with non-OIDC OAuth 2.0 providers, like GitHub, so we'd need to remove the id_token requirement at some point as well

File: crates/storage-pg/migrations/20241014145741_upstream_oauth_userinfo.sql
Comment (sandhose) [inline crates/storage-pg/migrations/20241014145741_upstream_oauth_userinfo.sql:7]:
```suggestion
  ADD COLUMN "userinfo" JSONB;
```

File: frontend/src/gql/graphql.ts
Comment (sandhose) [inline frontend/src/gql/graphql.ts:1]: I think you need to update your local dependencies in the `frontend/` folder

```sh
cd frontend/
npm ci
cd ..
sh misc/update.sh
```

File: crates/storage-pg/migrations/20241014145741_upstream_oauth_userinfo.sql
Comment (MatMaul) [reply crates/storage-pg/migrations/20241014145741_upstream_oauth_userinfo.sql:7]: This is going to be annoying : the return of `sqlx::query_as!` then wants a `serde_json::Value`.

But for example `data-model` crate doesn't depend on `serde_json`.

It may be fine to keep it a string ? There is currently not a lot of use of `JSONB` here, and it's not coherent with `id_token`.

Comment (sandhose) [reply crates/storage-pg/migrations/20241014145741_upstream_oauth_userinfo.sql:7]: I'm fine to depend on `serde_json` in `mas-data-model`. I'll be doing something similar in #3521 to store extra query parameters passed on the callback

File: crates/config/src/sections/upstream_oauth2.rs
Comment (sandhose) [inline crates/config/src/sections/upstream_oauth2.rs:430]: I'm not convinced we should replicate the Synapse behaviour here, to avoid implicit behaviour as much as possible.

What I would like to have instead, is just a boolean whether we do fetch the userinfo or not; this way, we always get/verify the ID token if we get one, we fetch the userinfo if it is configured to, and we make both available in the template, with the nice merge you did for the `user` property

File: crates/data-model/src/upstream_oauth2/session.rs
Comment (sandhose) [inline crates/data-model/src/upstream_oauth2/session.rs:22]: I'm fine with adding a dependency to `serde_json` here, make that a `serde_json::Value`

File: crates/handlers/src/upstream_oauth2/callback.rs
Comment (sandhose) [inline crates/handlers/src/upstream_oauth2/callback.rs:260]: I think I used the `environment` here the wrong way. In #3521, I made the environment 'static', and rather passed the dynamic context as template context. I'll let you know once that PR is merged so you can rebase your work on it

File: crates/handlers/src/upstream_oauth2/callback.rs
Comment (sandhose) [inline crates/handlers/src/upstream_oauth2/callback.rs:294]: This is better implemented with `userinfo.as_ref().map(|v| json!(v))`

File: crates/handlers/src/upstream_oauth2/callback.rs
Comment (sandhose) [inline crates/handlers/src/upstream_oauth2/callback.rs:253]: Interesting, I did not know we were checking that the userinfo returns the same sub & co as the id_token… we probably want to remove that, as we'd like to support non-OIDC providers as well…

I can do that in a later PR though

**Review by sandhose — CHANGES_REQUESTED** [review]
Skimming through this, it looks alright, thanks a lot! I think we need to wait for #3521 to land, so that you can rebase on top of it

File: crates/handlers/src/upstream_oauth2/callback.rs
Comment (sandhose) [reply crates/handlers/src/upstream_oauth2/callback.rs:260]: #3521 has been merged, so instead what I suggest is adding the `userinfo_claims` in the template context here: https://github.com/element-hq/matrix-authentication-service/blob/7296364cd3ea9abe06ae4bb67bffeb7fc28a5978/crates/handlers/src/upstream_oauth2/template.rs#L24

and then make sure it can be used in the template here: https://github.com/element-hq/matrix-authentication-service/blob/7296364cd3ea9abe06ae4bb67bffeb7fc28a5978/crates/handlers/src/upstream_oauth2/template.rs#L54-L79

**Review by sandhose — APPROVED** [review]
LGTM! It just occurred to me that it's missing documentation about the new options here: https://element-hq.github.io/matrix-authentication-service/reference/configuration.html#upstream_oauth2

But I'm fine with doing this in another PR

**Comment by pierreozoux** [comment]
Well done @MatMaul  :)

**Pull request merged**
